src/librustc_lint/array_into_iter.rs

```diff
@@ -0,0 +1,82 @@
+use crate::lint::{LateContext, LateLintPass, LintArray, LintContext, LintPass};
+use rustc::{
+    lint::FutureIncompatibleInfo,
+    hir,
+    ty,
+};
+use syntax::{
+    errors::Applicability,
+    symbol::{Symbol, sym},
+};
+
+
+declare_lint! {
+    pub ARRAY_INTO_ITER,
+    Warn,
+    "detects calling `into_iter` on arrays",
+    @future_incompatible = FutureIncompatibleInfo {
+        // FIXME: update with real issue
+        reference: "issue #0 <https://github.com/rust-lang/rust/issues/0>",
+        edition: None,
+    };
+}
+
+declare_lint_pass!(
+    /// Checks for instances of calling `into_iter` on arrays.
+    ArrayIntoIter => [ARRAY_INTO_ITER]
+);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for ArrayIntoIter {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx hir::Expr) {
+        // We only care about method call expressions.
+        if let hir::ExprKind::MethodCall(call, span, args) = &expr.kind {
+            if call.ident.name.as_str() != "into_iter" {
+                return;
+            }
+
+            // Test if the `self` type is an array type. Early return if not.
+            match cx.tables.expr_ty_adjusted(&args[0]).kind {
+                ty::Ref(_, ty::TyS { kind: ty::Array(..), .. }, _) => {}
```

Comment:
This doesn't handle the case of calling `into_iter` on arrays with more than 32 elements, where this would be `ty::Slice`. There probably needs to be some careful checking of the adjustments:

* In the slice case, if the last adjustment is not an `Adjust::Pointer(PointerCast::Unsize)` the don't lint.
* If there's fewer than 2 adjustments then lint. (fewer than 3 in the slice case)
* If the type of before the last 2 (3) adjustments is `&[T; N]` then don't lint.
* Otherwise, lint.

Reply:
Wait... really? Why is that? Sounds like a bad idea. Why are arrays <= 32 special cased? Could you provide a link to some docs where this is explained? The docs for `TyKind` does not mention this.

Reply:
All implementations on arrays that are provided by the standard library are limited to 32 elements (see `libcore/array/mod.rs`).

Reply:
Ahhh, ok, sorry, I misunderstood. I thought you were implying the compiler internally treats arrays >33 as slices. Which didn't make sense to me. But yeah, sure, due to a missing `impl IntoIterator for &[T; 33]`, it gets coerced to a slice. Makes sense now ^_^